### ISSUE TITLE
Do we care about code generation performance?

### ISSUE THREAD

**Issue by pavel-kirienko** [UAVCAN/nunavut#108]
It is not something that is done often so the benefit of such optimizations is low but it is non-zero. Experimentation may require one to re-generate code multiple times in a short period of time; doing so incrementally is currently impossible because we can only operate at the root namespace level (there are reasons for that related to data type compatibility, long story).

I wouldn't raise this issue if not for the fact that running a benign command like `pyuavcan -vv dsdl-gen-pkg ~/uavcan/public_regulated_data_types/uavcan/` takes 55 seconds, of which 51 appear to be used by Jinja.

See the performance profiles (output ordered by cumulative and total time): [perf.tar.gz](https://github.com/UAVCAN/nunavut/files/4464859/perf.tar.gz)

Are we OK with that or is there a cheap way to make Jinja faster? The performance issues were probably introduced when we disabled the cache to implement predictable side effects necessary for unique identifier generation:

https://github.com/thirtytwobits/nunavut/blame/7fd2ef1640430f746c8d0c00347a04fa2920d966/src/nunavut/jinja/__init__.py#L410

**Comment by thirtytwobits** [comment]
Yeah, the cache had to be disabled for the unique identifier feature. There was simply no other way to use that feature. We have three choices on that front:
1. do nothing
2. allow the cache to be re-enabled if that feature is not used.
3. remove the unique identifier generation feature.

**Comment by thirtytwobits** [comment]
This is causing considerable build delays. Try reimplementing unique identifier feature using something like this https://github.com/UAVCAN/pyuavcan/blob/908ba013072cbd297fe94615db7a417553d03c51/pyuavcan/dsdl/_templates/unique_reference.j2

**Comment by thirtytwobits** [comment]
@PetervdPerk-NXP : watch this space

**Comment by thirtytwobits** [comment]
Cache Disabled:
```
tox -e py39-test -- -k "test_realgen[c-True]"  76.19s user 0.73s system 99% cpu 1:17.04 total
```

Cache Enabled
```
tox -e py39-test -- -k "test_realgen[c-True]"  6.69s user 0.51s system 98% cpu 7.286 total
```

So...yes. This is the problem.
